### PR TITLE
Update brod_gssapi_v1 to use sasl_auth 2.0

### PR DESCRIPTION
**Pull request by starbelly**
- no longer define error codes and instead depend on atoms provided by sasl_auth 2.0
- bump version
- Remove call to list mechanisms as there isn't a need unless since we always want gssapi.

Co-authored-by: Vikas Bhardwaj <vikas15.bhardwaj@gmail.com>

# Notes 

- Depends on https://github.com/kafka4beam/sasl_auth/pull/8

- Changes were kept to a minimum, there's a lot of refactoring to be done in the v1 module though IMO, it's a bit hard to follow right now. 

- There are still no tests. We could do what was in done in sasl_auth v2 (i.e., tests that depend on a kdc and kafka instance). Long term some mock/fake servers would serve as well as far as being able to run simple integration tests. 

- It may be we want to just make the v1 module the default (i.e., just overwrite brod_gssapi.erl) since it won't even work with pulling in sasl_auth v2. It may be that we want to update that module to use sasl_auth v2, @vikas15bhardwaj Might have thoughts on this. 

- An improvement that could be made would be to put the list mechs call back in, but verify that gssapi is in there.

### CODE REVIEW

**Comment by vikas15bhardwaj** [comment]
thanks @starbelly for putting this out here.
On your point if we should make v1 default: here are my thoughts:

`brod_gssapi_v1` was added to support Kafka protocol version 1 and above. In my view, we should have one `brod_gssapi` plugin which should work for old Kafka protocols (older 0.0.9) or 1+. Users shouldn't be aware of this difference. Internally main plugin can bifurcate to either v1 or v0... depending on Kafka version it's dealing with. I think regardless of Kafka protocol `sasl_auth` version should be same

**Comment by starbelly** [comment]
> thanks @starbelly for putting this out here. On your point if we should make v1 default: here are my thoughts:
> 
> `brod_gssapi_v1` was added to support Kafka protocol version 1 and above. In my view, we should have one `brod_gssapi` plugin which should work for old Kafka protocols (older 0.0.9) or 1+. Users shouldn't be aware of this difference. Internally main plugin can bifurcate to either v1 or v0... depending on Kafka version it's dealing with. I think regardless of Kafka protocol `sasl_auth` version should be same

That makes sense. So we'll need to update the main plugin and test it.  

Edit:

I'll also meck out the nif calls and so forth so we can at least test the login in the modules.

**Comment by starbelly** [comment]
What's needed is the handshake vsn passed in from brod to the backend auth plugin. Should we send a patch upstream or not let perfect be the enemy of good here? 

I'm also taking a stab at refactoring both modules since they are hard to follow. I can tuck those changes to the side, but imo that can will just keep getting kicked down the road. 

@kjellwinblad @zmstone Any thoughts here?

File: src/brod_gssapi_v1.erl [hunk 1]
```diff
@@ -149,13 +99,20 @@ sasl_recv(Mod, Sock, Timeout, ClientId, Challenge, HandshakeVsn) ->
...
-    iolist_to_binary(Str).
-
-do_while(Fun, CondFun) ->
-    case CondFun(Fun()) of
+ensure_binary(Str) when is_list(Str) ->
+    iolist_to_binary(Str);
+ensure_binary(Bin) when is_binary(Bin) ->
+    Bin.
+
+maybe_interact_continue(sasl_interact) ->
```
Comment (starbelly) [inline src/brod_gssapi_v1.erl]: This needs to be removed all together. We should never hit `sasl_interact`. I think this was originally based off librdkafka it seems. librdkafka may have a case for interacting with a user (i.e., gui or cli app), but we have none or rather AFAIK we have no way of interacting with the user currently. 

When you provide callbacks to cyrus-sasl for additional information that may needed, you're opting out of getting `sasl_interact` back as a code AFAIK. What's more even if you do get that back at client step, you're supposed to go back to calling client start after you get the information asked for.

**Comment by kjellwinblad** [comment]
> What's needed is the handshake vsn passed in from brod to the backend auth plugin. Should we send a patch upstream or not let perfect be the enemy of good here?

Unfortunately, I have not been able to understand how this plugin works well enough yet to have an intelligent opinion. However, I think a patch to broad that sends the version to the plugin sounds like a good idea, if it is needed so that users don't need to configure what version to use manually. It is probably good if it is possible to do the change so other authentication plugins don't need to be changed. However, I'm not sure if that is a big deal since I'm don't know which other authentication plugins exists. @zmstone maybe you know more about this?

> I'm also taking a stab at refactoring both modules since they are hard to follow. I can tuck those changes to the side, but imo that can will just keep getting kicked down the road.

I think this sounds good as it is probably something that will pay off in the long run.

My current plan is to spend some more time with the `sasl_auth` repository to make sure it is production ready (maybe add some more test cases and study the code to make sure I can't find anything that is broken). After that, I plan to add test cases to this repo as well (probably by utilizing [docker-compose](https://docs.docker.com/compose/) to run brod, kafka and a Kerberos KDC in different Docker containers).

**Comment by zmstone** [comment]
> What's needed is the handshake vsn passed in from brod to the backend auth plugin. Should we send a patch upstream or not let perfect be the enemy of good here?

https://github.com/kafka4beam/kafka_protocol/blob/713a00e30a5f7ab0d19558c502a7e5cc5b0bfa68/src/kpro_sasl.erl#L36-L39
the version is not passed down to the callback, we can extend the callback to remain backward compatible.

**Comment by starbelly** [comment]
> Unfortunately, I have not been able to understand how this plugin works well enough yet to have an intelligent opinion.

@kjellwinblad  Hopefully what I've pushed up helps code wise, but need to add updated docs that covers the flow.  Interestingly, what I smell here is a behaviour where sasl_auth has a function that does the flow and depends on a few callbacks from the caller, but premature as well.








> https://github.com/kafka4beam/kafka_protocol/blob/713a00e30a5f7ab0d19558c502a7e5cc5b0bfa68/src/kpro_sasl.erl#L36-L39
> the version is not passed down to the callback, we can extend the callback to remain backward compatible.

@zmstone  I was thinking maybe adding a new callback function that passes in a map or record such that future changes don't require plugins to worry about different functions with different arity. See https://github.com/kafka4beam/brod_gssapi/pull/5/files#diff-6d37c08aa5bef5ce84cd840e06fe6c363227ed626e77af4010732b54948a1555R122 

I mainly did this to help clean up what's going on in this plugin, but then I figured I'd raise it as an option before I send up a patch to `kafka_protocol` 

Obviously, only brod_gssapi v2 and future backend auth plugins would utilize this. Thoughts?

**Comment by starbelly** [comment]
The changes at this point should require : https://github.com/kafka4beam/kafka_protocol/pull/102 .

Technically, we can skip that, and the auth/8 as part of the kafka_protocol PR, will just live doormant in the code until such time that it can be merged.

File: src/brod_gssapi.erl [hunk 2]
```diff
@@ -5,134 +5,110 @@
...
+        host => ensure_binary(Host),
+        sock => Sock,
+        transport_mod => Mod,
+        client_id => ClientId,
+        timeout => Timeout,
+        method => Method,
+        mechanism => <<"GSSAPI">>,
+        keytab => ensure_binary(KeyTab),
+        principal => ensure_binary(Principal),
+        sasl_context => <<"kafka">>,
```
Comment (kjellwinblad) [inline src/brod_gssapi.erl]: Can hard-coding this cause some problem?

File: src/brod_gssapi.erl [hunk 3]
```diff
@@ -5,134 +5,110 @@
...
+) ->
+    State = new(Host, Sock, HandshakeVsn, Mod, ClientId, Timeout, Opts),
+    dispatch(State).
 
-%%====================================================================
-%% Internal functions
-%%====================================================================
+dispatch(#{handshake_vsn := 1} = State) ->
+    brod_gssapi_v1:auth(State);
+dispatch(#{handshake_vsn := 0} = State) ->
```
Comment (kjellwinblad) [inline src/brod_gssapi.erl]: When is the the `handshake_vsn = 0`?

**Review by kjellwinblad — APPROVED** [review]
I think this looks very good!

**Comment by kjellwinblad** [comment]
I think I found a bug that I think should be fixed before merge (or we us the fix in PR https://github.com/kafka4beam/brod_gssapi/pull/6). See https://github.com/kafka4beam/brod_gssapi/pull/6#issuecomment-1150983984.

File: src/brod_gssapi.erl (hunk 2 above)
Comment (starbelly) [reply src/brod_gssapi.erl]: It should not be hard coded. We should use macros and/or functions.

File: src/brod_gssapi.erl (hunk 3 above)
Comment (starbelly) [reply src/brod_gssapi.erl]: This would be passed in from kafka_protocol, but it's probably  better as `undefined` since there is no handshake `0`.  Thanks for calling it out :) 

Edit:

I think that's wrong actually. So kafka_protocol will pass down either `0` or `1`. Kafka did not support a handshake at one point in time, which is what `brod_gssapi_v0` actually supports. Thus `brod_gssapi_v0` should really be called something else and `brod_gssapi_v1` should be adjusted to handle both v0 and v1 handshakes. See https://github.com/kafka4beam/kafka_protocol/blob/713a00e30a5f7ab0d19558c502a7e5cc5b0bfa68/src/kpro_sasl.erl#L75 

Also see https://kafka.apache.org/protocol.html#sasl_handshake 

and https://cwiki.apache.org/confluence/display/KAFKA/KIP-43%3A+Kafka+SASL+enhancements

It may be we don't want to support the non-handshake version, though brokers still do support it. Especially if kafka_protocol doesn't support it (i.e., kafka_protocol will always send 0 or 1 for the handshake version). 

Though, there's no harm in keeping it around as `brod_gssapi_legacy` for users who are using versions of kafka_protocol that don't pass the handshake in. 

Thoughts?

Comment (starbelly) [reply src/brod_gssapi.erl]: @zmstone Can you comment on the above? Either way, I'll have to get this merged tomorrow.

Comment (starbelly) [reply src/brod_gssapi.erl]: Actulally, here's what I'll do, I'll go ahead and merge this and open a few issues :)

Comment (starbelly) [reply src/brod_gssapi.erl]: https://github.com/kafka4beam/brod_gssapi/issues/8

Hated to merge it in the state it was in and with a few things being up in the air, but didn't want to block anymore + kaizen 😁

Comment (kjellwinblad) [reply src/brod_gssapi.erl]: I don't have any strong feeling either way. Perhaps @zmstone also wants to comment on this as he has much more experience with Kafka.
